components/javascript-wrapper.js

```diff
@@ -0,0 +1,157 @@
+       };
+
+    }  else if ( _.isUndefined( nodeDef.updater ) && !_.isFunction(nodeDef)) { 
+       // Got some node metadata, but no node updater - use the metadata with default updater
+       nodeDef = _.extend( nodeDef, 
+                           { updater: defaultUpdater } );
+
+    } else if ( _.isFunction( nodeDef ) ) { 
+        // Got an updater but no metadata - use updater args as input port list
+        // This assumes that all updater arguments are input
+        updaterArgs = introspect( nodeDef );
+        nodeDef = { inPorts: updaterArgs,
+                    updater: nodeDef };
+    } 
+
+    // Have an updater now, no matter what, be it default or not - figure out what the updater arguments are
+    var updaterArgs = (_.isNull( updaterArgs )) ? introspect( nodeDef.updater ) : updaterArgs; 
+    var inPorts = _.isArray(nodeDef.inPorts) ? _.object(nodeDef.inPorts, []) : nodeDef.inPorts;
+
+    return promiseComponent( 
+                         _.defaults({
+                             inPorts: _.mapObject(inPorts, function(port, name) {
+                                 return _.extend( port, 
+                                                  { ondata: ondata.bind( self, name, updaterArgs, nodeDef )});
+                             })
+                           }, nodeDef)
+                     );
+
+}; // module.exports
+
+// Define a default updater stub function to be used if the caller 
+// did not pass one in.  Simply returns whatever the original input was so it
+// can be sent through the output port to the next component
+var defaultUpdater = function() { 
+    return { out: arguments };
+}
+
+// Introspect the updater to determine what its parameters are.
+// The algorithm here comes from the following sources:
+//  - http://stackoverflow.com/questions/6921588/is-it-possible-to-reflect-the-arguments-of-a-javascript-function#answer-13660631
+//  - https://github.com/angular/angular.js/blob/master/src/auto/injector.js
+var introspect = function( fn ) {
+
+    var FN_ARGS = /^function\s*[^\(]*\(\s*([^\)]*)\)/m;
+    var FN_ARG_SPLIT = /,/;
+    var FN_ARG = /^\s*(_?)(\S+?)\1\s*$/;
+    var STRIP_COMMENTS = /((\/\/.*$)|(\/\*[\s\S]*?\*\/))/mg;
+  
+    try {
+        var fnText = fn.toString().replace(STRIP_COMMENTS, '');
+        var argDecl = fnText.match(FN_ARGS);
+        var rawArgs = argDecl[1].split(FN_ARG_SPLIT);
+
+        var args = []; // default to no args
+        if ( ! _.isEqual(rawArgs, [''] )) { 
+            // Got some args so extract them
+            args = _.map( rawArgs, function( arg ) { 
+                return arg.replace(FN_ARG, function(all, underscore, name) { return name });
+            });
+        }
+
+        return args;
+    } catch (e) { 
+        // Could throw exception if we do not have source to introspect
+        // Use default of empty argument list
+        return [];
+    }
+};
+
+var ondata = function( portName, updaterArgs, nodeDef, payload ) {
```

Comment:
I don't think the ondata function should be defined in a wrapper. It should be defined only once, as part of the underlying Framework layer.  The only API between the wrapper and the Framework should be through the three functions fExists, fRunUpdater and fDestroyState: https://github.com/rdf-pipeline/noflo-rdf-pipeline/wiki/Wrapper-API